Main.py

```diff
@@ -53,68 +53,78 @@ def __init__(self):
     def present_raw_data(self):
         plt.figure(1)
         T, = plt.plot(self.target[0, :])
-        F1, = plt.plot(self.feature1[0, :])
-        F2, = plt.plot(self.feature2[0, :])
-        F3, = plt.plot(self.feature3[0, :])
-        F4, = plt.plot(self.feature4[0, :])
+        F1, = plt.plot(self.feature[1 - 1][0, :])
+        F2, = plt.plot(self.feature[2 - 1][0, :])
+        F3, = plt.plot(self.feature[3 - 1][0, :])
+        F4, = plt.plot(self.feature[4 - 1][0, :])
         plt.legend([T, F1, F2, F3, F4], (self.DATA))
         plt.show(block=False)
 
     # preparing the data (Min max normalization and shape of the data)
-    # TODO split to normalization and reshaping
-    def prep_data(self):
-        self.X = np.concatenate([self.feature1, self.feature2, self.feature3, self.feature4])
+    def concatenate_data(self):
+        self.X = np.concatenate(self.feature)
         self.X = np.transpose(self.X)
 
         self.Y = self.target
         self.Y = np.transpose(self.Y)
 
-        scaler = MinMaxScaler()
+    def normalize(self):
+        scaler = MinMaxScaler(feature_range=(0, 1))
         scaler.fit(self.X)
         self.X = scaler.transform(self.X)
-        self.X = np.reshape(self.X, (self.X.shape[0], 1, self.X.shape[1]))
 
-        scaler1 = MinMaxScaler()
+        scaler1 = MinMaxScaler(feature_range=(0, 1))
         scaler1.fit(self.Y)
         self.Y = scaler1.transform(self.Y)
 
+    def reshape_X(self):
+        self.X = np.reshape(self.X, (self.X.shape[0], 1, self.X.shape[1]))
+        print(self.X.shape)
+
+    def prep_data(self):
+        self.concatenate_data()
+        self.normalize()
+        self.reshape_X()
+
     # split the data to train and test (as a batch)
     def split_train_test(self):
-        l = train_test_split(self.X, self.Y, test_size=0.3)
+        l = train_test_split(self.X, self.Y, test_size=0.2)
         self.X_train = l[0]
         self.X_test = l[1]
         self.Y_train = l[2]
         self.Y_test = l[3]
 
     # building and fitting the model
-    # TODO split to build function with parameters and train methods with parameters
-    def build_model(self):
+    def build_model(self, Nodes=50, LSTM_activation='tanh', recurrent_activation='hard_sigmoid',
+                    dense_activation='sigmoid', num_of_layers=1, optimizer='adam'):
         self.model = Sequential()
-        self.model.add(LSTM(100, activation='tanh', input_shape=(1, 4), recurrent_activation='hard_sigmoid'))
-        self.model.add(Dense(1))
-        self.model.compile(loss='mse', optimizer='rmsprop', metrics=[metrics.mae])
-        self.model.fit(self.X_train, self.Y_train, epochs=27, verbose=2)
+        for i in range(1, num_of_layers):
+            self.model.add(
+                LSTM(Nodes, activation=LSTM_activation, input_shape=(self.X.shape[1], self.X.shape[2]),
+                     recurrent_activation=recurrent_activation, return_sequences=True))
+        self.model.add(
```

Comment:
can you explain why we have another one here?

Reply:
This is a "hidden" LSTM layer, it helps the model to converge in less spochs in the price of more trainning time per epochs,
the default value is 1, which means that we never enter the ffor loop in line 114, but if you chose different value for "num_of_layers", you'll get more hidden layers.

it just another parameter i want to check.
i saw it on:[this video]( https://www.youtube.com/watch?v=Sqx2dmlJN6I)